src/librustc_driver/lib.rs

```diff
@@ -111,34 +111,48 @@ const BUG_REPORT_URL: &'static str = "https://github.com/rust-lang/rust/blob/mas
                                       md#bug-reports";
 
 #[inline]
-fn abort_msg(err_count: usize) -> String {
+fn abort_msg(err_count: usize) -> Option<String> {
     match err_count {
-        0 => "aborting with no errors (maybe a bug?)".to_owned(),
-        1 => "aborting due to previous error".to_owned(),
-        e => format!("aborting due to {} previous errors", e),
+        0 => Some("aborting with no errors (maybe a bug?)".to_owned()),
+        1 => None,
+        e => Some(format!("aborting due to {} previous errors", e)),
     }
 }
 
 pub fn abort_on_err<T>(result: Result<T, usize>, sess: &Session) -> T {
     match result {
         Err(err_count) => {
-            sess.fatal(&abort_msg(err_count));
+            if let Some(msg) = abort_msg(err_count) {
```

Comment:
maybe DRYer if abort_msg emits the error rather than returning a message, then the caller can just abort, rather than having to `if let` every time